File: 518-Coin-Change-II.cpp
class Solution {
public:
    int change(int amount, vector<int>& coins) {
        
        vector<long>dp(amount+1, 0);
        dp[0] =1;

        for(int coin : coins){
            for(int i =coin;i<=amount;i++){
                if(dp[i]<INT_MAX && dp[i-coin]<INT_MAX){
                    dp[i] +=dp[i-coin];

                }
            }
        }

        return dp[amount];

    }
};

File: 518-Coin-Change-II.py
class Solution:
    def change(self, amount: int, coins: List[int]) -> int:

        def solve(i, amount, dp):

            if amount==0:
                return 1

            if i==len(coins) or amount<0:
                return 0

            if dp[i][amount]!=-1:
                return dp[i][amount]

            take = solve(i, amount-coins[i],dp)
            notTake = solve(i+1, amount, dp)

            dp[i][amount] = take+notTake
            return dp[i][amount]

        dp = [[-1]*(amount+1) for _ in range(len(coins))]

        return solve(0, amount, dp)